quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,584 @@
+    }
+
+    /// Remove `timer` from its list
+    fn list_unlink(&mut self, timer: Timer) {
+        let prev = self.timers[timer.0].prev.take();
+        let next = self.timers[timer.0].next.take();
+        if let Some(prev) = prev {
+            // Remove reference from predecessor
+            self.timers[prev.0].next = next;
+        }
+        if let Some(next) = next {
+            // Remove reference from successor
+            self.timers[next.0].prev = prev;
+        }
+    }
+}
+
+fn range_in_level(level: usize, raw: RangeInclusive<u64>) -> RangeInclusive<usize> {
+    let shift = level * LOG_2_SLOTS;
+    const MASK: u64 = SLOTS as u64 - 1;
+    let start = ((*raw.start() >> shift) & MASK) as usize;
+    let level_end = (*raw.start() >> shift) | MASK;
+    let end = ((*raw.end() >> shift).min(level_end) & MASK) as usize;
+    start..=end
+}
+
+/// Compute the first tick that lies within a slot
+fn slot_start(base: u64, level: usize, slot: usize) -> u64 {
+    let shift = (level * LOG_2_SLOTS) as u64;
+    // Shifting twice avoids an overflow when level = 10.
+    (base & ((!0 << shift) << LOG_2_SLOTS as u64)) | ((slot as u64) << shift)
+}
+
+/// Compute the level and slot for a certain expiry
+fn timer_index(base: u64, expiry: u64) -> (usize, usize) {
+    // The level is the position of the first bit set in `expiry` but not in `base`, divided by the
+    // number of bits spanned by each level.
+    let differing_bits = base ^ expiry;
+    let level = (63 - (differing_bits | 1).leading_zeros()) as usize / LOG_2_SLOTS;
+    debug_assert!(level < LEVELS, "every possible expiry is in range");
+
+    // The slot in that level is the difference between the expiry time and the time at which the
+    // level's span begins, after both times are shifted down to the level's granularity. Each
+    // level's spans starts at `base`, rounded down to a multiple of the size of its span.
+    let slot_base = (base >> (level * LOG_2_SLOTS)) & (!0 << LOG_2_SLOTS);
+    let slot = (expiry >> (level * LOG_2_SLOTS)) - slot_base;
+    debug_assert!(slot < SLOTS as u64);
+
+    (level, slot as usize)
+}
+
+impl<T> Default for DelayQueue<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+struct TimerState<T> {
+    /// Lowest argument to `poll` for which this timer may be returned
+    expiry: u64,
+    /// Value returned to the caller on expiry
+    value: T,
+    /// Predecessor within a slot's list
+    prev: Option<Timer>,
+    /// Successor within a slot's list
+    next: Option<Timer>,
+}
+
+/// A set of contiguous timer lists, ordered by expiry
+#[derive(Copy, Clone)]
+struct Level {
+    slots: [Option<Timer>; SLOTS],
+}
+
+impl Level {
+    fn new() -> Self {
+        Self {
+            slots: [None; SLOTS],
+        }
+    }
+}
+
+impl fmt::Debug for Level {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut m = f.debug_map();
+        for (i, Timer(t)) in self
```

Comment:
How do you feel about naming the scrutinee (maybe just `iter`) to avoid the multi-line scrutinee -- which makes the block header harder to follow?

Reply:
Sure, that does seem a bit clearer.